polar-core/src/filter.rs

```diff
@@ -0,0 +1,491 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::{self, Display, Formatter},
+    hash::Hash,
+};
+
+use crate::{
+    data_filtering::{unregistered_field_error, unsupported_op_error, PartialResults, Type, Types},
+    error::{invalid_state_error, PolarResult, RuntimeError},
+    events::ResultEvent,
+    terms::*,
+};
+
+use serde::Serialize;
+
+type TypeName = String;
+type FieldName = String;
+type Map<A, B> = HashMap<A, B>;
+type Set<A> = HashSet<A>;
+
+#[derive(Clone, Eq, Debug, Serialize, PartialEq)]
+pub struct Filter {
+    root: TypeName,                  // the host already has this, so we could leave it off
+    conditions: Vec<Set<Condition>>, // disjunctive normal form
+    relations: Set<Relation>,
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Clone, Hash)]
+pub struct Relation(TypeName, FieldName, TypeName);
+
+#[derive(PartialEq, Eq, Debug, Serialize, Clone, Hash)]
+pub struct Condition(Datum, Compare, Datum);
+
+#[derive(PartialEq, Eq, Debug, Serialize, Clone, Hash)]
+pub enum Datum {
+    Field(Proj),
+    Imm(Value),
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Clone, Hash)]
+pub struct Proj(TypeName, Option<FieldName>);
+
+#[derive(PartialEq, Debug, Serialize, Copy, Clone, Eq, Hash)]
+pub enum Compare {
+    Eq,
+    Neq,
+    In,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+struct PathVar {
+    var: String,
+    path: Vec<String>,
+}
+
+impl From<String> for PathVar {
+    fn from(var: String) -> Self {
+        Self { var, path: vec![] }
+    }
+}
+
+impl From<Proj> for PathVar {
+    fn from(Proj(var, field): Proj) -> Self {
+        let path = field.into_iter().collect();
+        PathVar { var, path }
+    }
+}
+
+impl PathVar {
+    fn from_term(t: &Term) -> PolarResult<Self> {
+        use Value::*;
+        match t.value() {
+            Expression(Operation {
+                operator: Operator::Dot,
+                args,
+            }) => {
+                let dot = args[1].value().as_string()?.to_string();
+                let mut pv = Self::from_term(&args[0])?;
+                pv.path.push(dot);
+                Ok(pv)
+            }
+            Variable(Symbol(var)) => Ok(var.clone().into()),
+            _ => invalid_state_error(format!("PathVar::from_term({})", t.to_polar())),
+        }
+    }
+}
+
+impl Operation {
+    /// turn an isa from the partial results into a pathvar -> type pair
+    fn into_entity(self) -> Option<PolarResult<(PathVar, TypeName)>> {
+        match self.args[1].value() {
+            Value::Pattern(Pattern::Instance(InstanceLiteral { tag, fields }))
+                if fields.is_empty() =>
+            {
+                // FIXME(gw) this is to work around a complicated simplifier
```

Comment:
That's correct. It's not fixed yet, but the host would just go through the partials, pick out the isas with an instance on the left, and either drop the constraint (if it's true) or drop the result set it appears in (if it's false)

Reply:
Just thinking about this more... shouldn't we be able to check this isa by using the type map?

Reply:
I guess not always.

Reply:
Either way, I'm wondering whether we should have this error rather than swallow this until we implement checking for this condition in the host library.

If we don't check these isa constraints, can we have incorrect query results?